### ISSUE TITLE
pg-native - How to bundle?

### ISSUE THREAD

**Issue by rentrop** [elastic-coders/serverless-webpack#78]
Importing `pg.Client` via `var Client = require('pg').Client;` throws the following error:

> Module not found: Error: Cannot resolve module 'pg-native' in .../node_modules/pg/lib/native
 @ ./~/pg/lib/native/index.js 9:13-33

How can i fix this? I found the following somewhat related Issues/answers

1. https://github.com/serverless/serverless-runtime-babel/issues/8 by @thenikso 
2. https://github.com/brianc/node-postgres/issues/1187

System-Info:

__.babelrc__
{
  "plugins": ["transform-runtime"],
  "presets": ["es2015", "stage-0"],
  "ignore": [
   "node_modules/pg/lib/native/*",
   "node_modules/pg/lib/native/index.js",
   "node_modules/pg/lib/native/result.js",
   "node_modules/pg/lib/native/query.js",
   "./node_modules/pg/lib/native/*",
   "./node_modules/pg/lib/native/index.js",
   "./node_modules/pg/lib/native/result.js",
   "./node_modules/pg/lib/native/query.js",
   "~/pg/lib/native/index.js",
   "node_modules/pg/**",
   "node_modules/pg/*",
   "./node_modules/pg/",
   "./node_modules/pg/*",
   "./node_modules/pg/**"
  ]
}

__webpack.config.js__
Sames as in https://github.com/elastic-coders/serverless-webpack/blob/master/examples/babel/webpack.config.js

```
  Your Environment Information -----------------------------
     OS:                 darwin
     Node Version:       4.3.2
     Serverless Version: 1.5.0
```

**Comment by thenikso** [comment]
I'll soon work on a peace of software which uses `pg-promise` that also imports `pg-native`. I fixed this some time ago but it probably needs refinement. As it was I had this in the `resolve` section of my `webpack.config.js`:

```
   alias: {
      'pg-native': path.join(__dirname, 'aliases/pg-native.js'),
      'pgpass$': path.join(__dirname, 'aliases/pgpass.js'),
    },
```

And the files are:

```
// alias/pg-native.js
export default null;
```

```
// alias/pgpass.js
'use strict';

var fs = require('fs')
  , helper = require( 'pgpass/lib/helper.js' )
;


module.exports.warnTo = helper.warnTo;

module.exports = function(connInfo, cb) {
    var file = helper.getFileName();

    fs.stat(file, function(err, stat){
        if (err || !helper.usePgPass(stat, file)) {
            return cb(undefined);
        }

        var st = fs.createReadStream(file);

        helper.getPassword(connInfo, st, cb);
    });
};
```

**Comment by rentrop** [comment]
Perfect, thanks.

**Comment by vitaly-t** [comment]
While pg-promise fully supports pg-native, the real performance benefit is negligible, especially under the new versions of Node.js

**Comment by adieuadieu** [comment]
Another solution (if using webpack), is to add `new webpack.IgnorePlugin(/^pg-native$/)` to your webpack config's plugins array. E.g.

```
const webpackConfig = {
  ...
  resolve: { ... },
  plugins: [
    new webpack.IgnorePlugin(/^pg-native$/)
  ]
  output: { ... }
  ...
}
```

**Comment by aychtang** [comment]
For those experiencing this using serverless-bundle, and not wanting to migrate to serverless-webpack for their project, I have made this package for now: "serverless-bundle-no-pg-native": "1.3.0". This will allow you to deploy your lambda function using pg by swapping the serverless-bundle plugin to `serverless-bundle-no-pg-native`

You can check the diff at https://github.com/aychtang/serverless-bundle. I will not be maintaining this so it will stay at the current version of serverless-bundle, but I thought it'd be useful to share if people are doing some small projects and want to save some time.

TLDR:

`npm i serverless-bundle-no-pg-native --save-dev`.
Change plugin used in serverless.yml to `serverless-bundle-no-pg-native`.

**Comment by databyte** [comment]
Just came across this, if you're using `serverless-bundle`, add this to your serverless.yml

```yml
custom:
  bundle:
    ignorePackages:
      - pg-native
```

See: https://www.npmjs.com/package/serverless-bundle#pg

**Comment by joinnusdev** [comment]
> Just came across this, if you're using `serverless-bundle`, add this to your serverless.yml
> 
> ```yaml
> custom:
>   bundle:
>     ignorePackages:
>       - pg-native
> ```
> 
> See: https://www.npmjs.com/package/serverless-bundle#pg

that works for me, thanks

**Comment by getkey** [comment]
If you are using Webpack 5 the API of `IgnorePlugin` changed so [@adieuadieu's code](https://github.com/serverless-heaven/serverless-webpack/issues/78#issuecomment-405646040) needs a little tweak.

```js
const { IgnorePlugin } = require('webpack');

module.exports = {
	plugins: [
		new IgnorePlugin({
			resourceRegExp: /^pg-native$/,
		}),
	],
};
```

**Comment by Hacaw** [comment]
> If you are using Webpack 5 the API of `IgnorePlugin` changed so [@adieuadieu's code](https://github.com/serverless-heaven/serverless-webpack/issues/78#issuecomment-405646040) needs a little tweak.
> 
> ```js
> const { IgnorePlugin } = require('webpack');
> 
> module.exports = {
> 	plugins: [
> 		new IgnorePlugin({
> 			resourceRegExp: /^pg-native$/,
> 		}),
> 	],
> };
> ```

Thanks @getkey. It worked !

**Comment by bibhuticoder** [comment]
> Just came across this, if you're using `serverless-bundle`, add this to your serverless.yml
> 
> ```yaml
> custom:
>   bundle:
>     ignorePackages:
>       - pg-native
> ```
> 
> See: https://www.npmjs.com/package/serverless-bundle#pg

Thanks. Saved my day

**Comment by yuleicul** [comment]
If you are using `serverless-esbuild`, add this to your serverless.yml:

```yml
custom:
  esbuild:
    exclude:
      - pg-native
```
See: https://www.npmjs.com/package/serverless-esbuild#configuration

**Comment by alexporrello** [comment]
> If you are using Webpack 5 the API of `IgnorePlugin` changed so [@adieuadieu's code](https://github.com/serverless-heaven/serverless-webpack/issues/78#issuecomment-405646040) needs a little tweak.
> 
> ```js
> const { IgnorePlugin } = require('webpack');
> 
> module.exports = {
> 	plugins: [
> 		new IgnorePlugin({
> 			resourceRegExp: /^pg-native$/,
> 		}),
> 	],
> };
> ```

You're the best! Thank you!
